Added takeScreenshot Action

mehsaandev opened the pull request.

kmahmood74 commented:
@mehsaandev no reviewer?

kmahmood74 commented:
also which ticket does this address? add the proposed yaml to that ticket. Let's get things done properly please

mehsaandev commented:
@snehmehta PTAL

kmahmood74 commented on modules/ensemble/lib/action/take_screenshot.dart:
```diff
@@ -0,0 +1,111 @@
+          event: EnsembleEvent(initiator, data: {'error': e.toString()}),
+        );
+      }
+      rethrow;
```
have you actually tested it? rethrowing the error may cause issues. TEST the error case please

mehsaandev commented on modules/ensemble/lib/action/take_screenshot.dart:
```diff
@@ -0,0 +1,111 @@
+        throw LanguageError("Widget not found: '$widgetId'");
+      }
+
+      final widget = Screenshot(
```
Yes it is only limited to read-only widgets, it won't work for Input widgets. I'll update this to capture Input widgets as well but not sure if the screenshot will contain the value of any Input widget.

mehsaandev commented on modules/ensemble/lib/action/take_screenshot.dart:
```diff
@@ -0,0 +1,111 @@
+      final widget = Screenshot(
+        controller: screenshotController,
+        child: DataScopeWidget(
+          scopeManager: scopeManager,
```
Taking screenshot from bottom sheet and using widgetId of the widget in the screen works fine
And it also works for dialog as well.

mehsaandev commented on modules/ensemble/lib/action/take_screenshot.dart:
```diff
@@ -0,0 +1,111 @@
+            child: widget,
+          ),
+        ),
+        delay: const Duration(milliseconds: 1000),
```
Basically in the package it was written that optimal delay is 1s so that's why I put 1000ms. It was recommended by package

kmahmood74 reviewed (changes requested):
1. add comments everywhere in the code wherever you are making assumptions. 
2. test the action in js by calling it as `ensemble.takeScreenshot`

once done let me know

mehsaandev commented:
> 1. add comments everywhere in the code wherever you are making assumptions.
> 2. test the action in js by calling it as `ensemble.takeScreenshot`
> 
> once done let me know

- Added necessary comments
- previously I was not saving the screenshot in gallery because takeScreenshot action returning imageBytes in response and we can further save it using `saveFile` action but `ensemble.takeScreenshot` can't return anything and it made sense to save the screenshot so I reused the function from saveFile action and saved the screenshot when we hit `takeScreenshot`
- Tested using `ensemble.takeScreenshot` and it's working fine
- Removed delay parameter because `captureFromLongWidget` method has default delay of 1s already
- Made a separate file for file saving function in `saveFile` action so that I can reuse them in my `takeScreenshot` action
